# METAGENOTE

## Synopsis ## 
METAGENOTE is a project from the National Institutes of Health that aims to aid researchers in applying standardized metadata describing the “what”, “where”, “how”, and “when” of the samples collected in a genomics study. The name METAGENOTE refers to the collection of METAdata of GEnomics studies in a web based NOTEbook. The metadata gets stored in a centralized repository and validated according to guidelines from the Genomics Standard Consortium, which are also supported by repositories and large microbiome initiatives such as NCBI, European Bioinformatics Institute (EBI), and Earth Microbiome Project. Upon request from the researchers, the data will also be submitted for publication via the NCBI Sequence Read Archive (SRA) repository.

>>Important note: As of May 1st, 2025, this project will no longer be available for users via the webpage https://metagenote.niaid.nih.gov 

To learn more about METAGENOTE, please see citation below:
Quiñones, M., Liou, D.T., Shyu, C. et al. METAGENOTE: a simplified web platform for metadata annotation of genomic samples and streamlined submission to NCBI’s sequence read archive. BMC Bioinformatics 21, 378 (2020). https://doi.org/10.1186/s12859-020-03694-0

### Relevant resources for metadata annotation, ontologies and submission
-	[NCBI SRA](https://submit.ncbi.nlm.nih.gov/about/sra/): The SRA is NIH's primary archive of high-throughput sequencing data and is part of the International Nucleotide Sequence Database Collaboration (INSDC) that includes at the NCBI Sequence Read Archive (SRA), the European Bioinformatics Institute (EBI), and the DNA Database of Japan (DDBJ). Data submitted to any of the three organizations are shared among them.  
-	[Genomics Standards Consortium](https://www.gensc.org/): The Genomic Standards Consortium (GSC) is an initiative working towards richer descriptions of our collection of genomes, metagenomes and marker genes. Established in September 2005, this international community includes representatives from a range of major sequencing and bioinformatics centres (including NCBI, EMBL, DDBJ, JCVI, JGI, EBI, Sanger, FIG) and research institutions.
-	[Ontology Lookup Service (EMBL)](https://www.ebi.ac.uk/ols4/index): The Ontology Lookup Service (OLS) is a repository for biomedical ontologies that aims to provide a single point of access to the latest ontology versions. You can browse the ontologies through the website as well as programmatically via the OLS API. OLS is developed and maintained by the Samples, Phenotypes and Ontologies Team at EMBL-EBI.
-	[NMDC Submission Portal and Field Notes App](https://microbiomedata.org/field-notes/): Tools developed by the National Microbiome Data Collaborative for capturing metadata and adhering to community standards.  
